### PR TITLE
Fix `LOGICAL_ERROR` if fetch of projection was stopped

### LINKED ISSUE #44054: Logical error: 'Cannot commit shared transaction'
https://s3.amazonaws.com/clickhouse-test-reports/43986/7249ad407a0824d8c7651d8e03d4e68e1dea0813/stress_test__msan_.html

### PR DESCRIPTION
**Pull request by CurtizJ**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix `LOGICAL_ERROR` in case when fetch of part was stopped while fetching projection to the disk with enabled zero-copy replication.

Fixes #44054.

### CODE REVIEW

**Comment by CurtizJ** [comment]
Bugfix validate check: see comment in test.

**Review by kssenii — APPROVED** [review]

**Comment by CurtizJ** [comment]
Stateless tests (debug) - at least one similar failure in master: https://s3.amazonaws.com/clickhouse-test-reports/0/bca18298a14bbc311342d757878aea631f16f8df/stateless_tests__debug__[2/3].html

**Comment by CurtizJ** [comment]
`02344_describe_cache.sql` - became flaky but it still passes with previous version.
`02494_zero_copy_projection_cancel_fetch.sh` - should fail only in debug build.

File: src/Storages/MergeTree/DataPartsExchange.cpp [hunk 1]
```diff
@@ -934,22 +931,36 @@ MergeTreeData::MutableDataPartPtr Fetcher::downloadPartToDisk(
...
+    catch (const Exception & e)
+    {
+        /// Remove the whole part directory if fetch of base
+        /// part or fetch of any projection was stopped.
+        if (e.code() == ErrorCodes::ABORTED)
+        {
+            data_part_storage->removeRecursive();
+            data_part_storage->commitTransaction();
+        }
+        throw;
```
Comment (tavplubix) [inline src/Storages/MergeTree/DataPartsExchange.cpp:962]: What about other error codes? See https://github.com/ClickHouse/ClickHouse/issues/54446

Comment (CurtizJ) [reply src/Storages/MergeTree/DataPartsExchange.cpp:962]: This PR just restores the previous behaviour with one small fix for projections. This behaviour existed all the time. Probably we rely on background thread for deleting of temp directories.

**Pull request merged**